py_ecc/bls_api/utils.py

```diff
@@ -0,0 +1,228 @@
+        x2 = -x1
+        x1_re, x1_im = x1.coeffs
+        x2_re, x2_im = x2.coeffs
+        return x1 if (x1_im > x2_im or (x1_im == x2_im and x1_re > x2_re)) else x2
+    return None
+
+
+def _get_x_coordinate(message_hash: Hash32, domain: int) -> FQ2:
+    domain_in_bytes = domain.to_bytes(8, 'big')
+
+    # Initial candidate x coordinate
+    x_re = big_endian_to_int(hash_eth2(message_hash + domain_in_bytes + b'\x01'))
+    x_im = big_endian_to_int(hash_eth2(message_hash + domain_in_bytes + b'\x02'))
+    x_coordinate = FQ2([x_re, x_im])  # x_re + x_im * i
+
+    return x_coordinate
+
+
+def hash_to_G2(message_hash: Hash32, domain: int) -> G2Uncompressed:
+    x_coordinate = _get_x_coordinate(message_hash, domain)
+
+    # Test candidate y coordinates until a one is found
+    while 1:
+        y_coordinate_squared = x_coordinate ** 3 + FQ2([4, 4])  # The curve is y^2 = x^3 + 4(i + 1)
+        y_coordinate = modular_squareroot(y_coordinate_squared)
+        if y_coordinate is not None:  # Check if quadratic residue found
+            break
+        x_coordinate += FQ2([1, 0])  # Add 1 and try again
+
+    return multiply(
+        G2Uncompressed((x_coordinate, y_coordinate, FQ2([1, 0]))),
+        G2_cofactor,
+    )
+
+
+#
+# G1
+#
+def compress_G1(pt: G1Uncompressed) -> G1Compressed:
+    """
+    A compressed point in group G1 is 384-bit integer with the bit order (c_flag, b_flag, a_flag, x),
+    where the c_flag bit is always set to 1,
+    the b_flag bit indicates infinity when set to 1,
+    the a_flag bit helps determine the y-coordinate when decompressing,
+    and the 381-bit integer x is the x-coordinate of the point.
+    """
+    if is_inf(pt):
+        # Set c_flag = 1 and b_flag = 1. leave a_flag = x = 0
+        return G1Compressed(POW_2_383 + POW_2_382)
+    else:
+        x, y = normalize(pt)
+        # Record y's leftmost bit to the a_flag
+        a_flag = (y.n * 2) // q
+        # Set c_flag = 1 and b_flag = 0
+        return G1Compressed(x.n + a_flag * POW_2_381 + POW_2_383)
+
+
+def decompress_G1(z: G1Compressed) -> G1Uncompressed:
+    """
+    Recovers x and y coordinates from the compressed point.
+    """
+    # b_flag == 1 indicates the infinity point
+    b_flag = (z % POW_2_383) // POW_2_382
+    if b_flag == 1:
+        return G1Uncompressed(Z1)
+    x = z % POW_2_381
+
+    # Try solving y coordinate from the equation Y^2 = X^3 + b
+    # using quadratic residue
+    y = pow((x**3 + b.n) % q, (q + 1) // 4, q)
+
+    if pow(y, 2, q) != (x**3 + b.n) % q:
+        raise ValueError(
+            "The given point is not on G1: y**2 = x**3 + b"
+        )
+    # Choose the y whose leftmost bit is equal to the a_flag
+    a_flag = (z % POW_2_382) // POW_2_381
+    if (y * 2) // q != a_flag:
+        y = q - y
+    return G1Uncompressed((FQ(x), FQ(y), FQ(1)))
+
+
+def G1_to_pubkey(pt: G1Uncompressed) -> BLSPubkey:
+    z = compress_G1(pt)
```

Comment:
Moved compression and decompression to the `G1_to_pubkey`, `pubkey_to_G1`, `G2_to_signature`, and `signature_to_G2` to prettify the function bodies of functions in the api.py.